### PR TITLE
IIIF canvas dimension task

### PR DESCRIPTION
**Pull request by mspalti**
## References
* https://github.com/DSpace/dspace-angular/issues/1396 (the issue should have been created for dspace, not dspace-angular)

## Description
Adds a new DSpace cli task for creating canvas dimension bitstream metadata.  It adds metadata to image bitstreams that belong to an IIIIF enabled item and are found in an IIIF bundle. Requires an identifier for a community, collection or item as a starting point. Additional flags for force processing, maximum items to process, skip list, and quiet mode.

Example:
```
./dspace iiif-canvas-dimensions -e user@email.org -i 1086306d-8a51-43c3-98b9-c3b00f49105f -t COLLECTION
```

## Instructions for Reviewers

List of changes in this PR:
* Updated launcher.xml
* "iiif-processing" spring xml config
* New IIIF utils class (shared with dspace-iiif package)
* New "canvasdimension" package for processing
* Integration tests

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [ ] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [ ] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

### CODE REVIEW

File: dspace-api/src/main/java/org/dspace/app/canvasdimension/IIIFCanvasDimensionServiceImpl.java
Comment (abollini) [inline dspace-api/src/main/java/org/dspace/app/canvasdimension/IIIFCanvasDimensionServiceImpl.java:154]: it would be better imho to try to read any image and catch eventually the exceptions/null results

File: dspace-api/src/main/java/org/dspace/app/canvasdimension/IIIFCanvasDimensionServiceImpl.java
Comment (abollini) [inline dspace-api/src/main/java/org/dspace/app/canvasdimension/IIIFCanvasDimensionServiceImpl.java:169]: the inputstream must be closed in a finally block

File: dspace-api/src/main/java/org/dspace/app/canvasdimension/IIIFCanvasDimensionServiceImpl.java
Comment (abollini) [inline dspace-api/src/main/java/org/dspace/app/canvasdimension/IIIFCanvasDimensionServiceImpl.java:192]: metadata should be added with `null` for the language, it is fine to clear them with ANY to cleanup manually entered information

File: dspace-api/src/main/java/org/dspace/app/canvasdimension/CanvasDimensionCLI.java
Comment (abollini) [inline dspace-api/src/main/java/org/dspace/app/canvasdimension/CanvasDimensionCLI.java:55]: this is not strictly needed. The script has a value also when iiif is not enabled because it adds technical metadata at the bitstream level. It is true that without IIIF the API service will be unable to return any information but this is just the fallback for the scenario where the ImageIO class is not able to calculate it.
I would suggest to just print a warning message instead than refuse to run the script

File: dspace-api/src/main/java/org/dspace/app/canvasdimension/IIIFApiQueryServiceImpl.java
Comment (abollini) [inline dspace-api/src/main/java/org/dspace/app/canvasdimension/IIIFApiQueryServiceImpl.java:76]: as in https://github.com/DSpace/DSpace/pull/8047/files#r783792920
this must be logged

File: dspace-api/src/main/java/org/dspace/app/canvasdimension/IIIFApiQueryServiceImpl.java
Comment (abollini) [inline dspace-api/src/main/java/org/dspace/app/canvasdimension/IIIFApiQueryServiceImpl.java:70]: as in https://github.com/DSpace/DSpace/pull/8047/files#r783791480
the input stream should be closed in a finally block to avoid connection leaks in case of exception

File: dspace-api/src/main/java/org/dspace/app/canvasdimension/IIIFApiQueryServiceImpl.java
Comment (abollini) [inline dspace-api/src/main/java/org/dspace/app/canvasdimension/IIIFApiQueryServiceImpl.java:57]: as in https://github.com/DSpace/DSpace/pull/8047/files#r783791480
would be possible to delegate the generation of the info.json url to the `ImageContentService` ?

**Review by abollini — CHANGES_REQUESTED** [review]
Hi @mspalti thanks for this valuable contribution.
It is indeed a feature that was provided by the 4Science's IIIF Addon that is very appreciated by customer. It is a pity that we were not aware of your intention to work on it otherwise we would have considered the option to donate what we already have using it as starting point to work together and improve the solution.  That said, your solution looks great I have included some suggestion inline to further improve it according to our usage experience

File: dspace-api/src/main/java/org/dspace/app/canvasdimension/IIIFCanvasDimensionServiceImpl.java
Comment (mspalti) [reply dspace-api/src/main/java/org/dspace/app/canvasdimension/IIIFCanvasDimensionServiceImpl.java:196]: Hi @abollini , so `IIIFUtils` and `ImageContentService` are part of the spring boot web application. Since we already have an `IIIFSharedUtils` class in `org.dspace.iiif` it might make sense for both the cli process and web application to use the shared utils for constants and the `info.json` path. I'll refactor accordingly. Let me know if you see a better approach.

File: dspace-api/src/main/java/org/dspace/app/canvasdimension/Util.java
Comment (mspalti) [reply dspace-api/src/main/java/org/dspace/app/canvasdimension/Util.java:27]: It's poorly worded and  a bit confusing but the comment I copied from the IIIF specification says less than 1200 pixels on either edge, which seems to say we need ` ||`. I will double check the spec.

File: dspace-api/src/main/java/org/dspace/app/canvasdimension/IIIFApiQueryServiceImpl.java
Comment (mspalti) [reply dspace-api/src/main/java/org/dspace/app/canvasdimension/IIIFApiQueryServiceImpl.java]: I added the method to `org.dspace.iiif.IIIFSharedUtils`.  On the IIIF API side we can use this method from within `ImageContentService`.

**Comment by mspalti** [comment]
@abollini , I wish we had started working together on this one, too. I knew that I had jumped into work that might be helped along by more discussion in the community and 4Science input. There were just some things that really needed to be done within the timeline for 7.2 and that's the reason for the quick start. Thanks for your willingness to collaborate!

**Comment by mspalti** [comment]
@abollini , if you approve of the changes I made here, I will address your similar suggestions for #8015 after this PR is merged. They can both rely on the new `IIIFSharedUtils`

File: dspace-api/src/main/java/org/dspace/app/canvasdimension/CanvasDimensionCLI.java
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/app/canvasdimension/CanvasDimensionCLI.java:119]: Can we print out the help text here as well?  The first time I ran this with no arguments, I expected it to error out, but also tell me what a valid argument would look like.  

There's also no "--help" or "-h" option...so, it's impossible to see the list of valid arguments from commandline.  We should add a "-h" option

File: dspace/config/launcher.xml
Comment (tdonohue) [inline dspace/config/launcher.xml:371]: We may want to improve this description, as I first understood it to say that it adds the canvas height/width __to the Item__. But, it's obviously adding it to the Bitstream.  Maybe say something like this?

"Automatically add the IIIF canvas width/height metadata to Bitstreams of IIIF enabled Items".

File: dspace-api/src/main/java/org/dspace/app/canvasdimension/CanvasDimensionCLI.java
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/app/canvasdimension/CanvasDimensionCLI.java:192]: We might want to log an error of some sort if an unsupported object type is passed in. Otherwise, it appears this will just "do nothing" and be quiet about it.

File: dspace-api/src/main/java/org/dspace/app/canvasdimension/CanvasDimensionCLI.java
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/app/canvasdimension/CanvasDimensionCLI.java:195]: We should only commit changes if something is updated.  this should at least check that >=1 items were processed

**Review by tdonohue — CHANGES_REQUESTED** [review]
@mspalti : I tested this today & also gave it a quick code review. With regards to testing, it mostly works, but it's a bit hard to use for a few reasons:

1. UUIDs don't appear to work.  So, an argument `./dspace canvas-dimensions -i [uuid] -e [email]` results in an error: `Cannot resolve [uuid] to a DSpace object`.  Only Handles work.  Ideally either UUIDs or handles would work if possible.
2. The CLI command is missing help information, i.e. an `-h` option should exist, otherwise you cannot figure out what commands are valid without looking at documentation (or code), see more inline below.

Some other comments inline regarding the code structure and other minor suggestions.

**Comment by mspalti** [comment]
@tdonohue , I think that using UUID as the identifier means we require an additional parameter that specifies the dso type (Community, Collection, Item). That's what I see in other scripts, like `ItemExportCLITool`.  Basically we need to know which service to call so we can do the UUID lookup. I'm adding this now.

This script currently works like media filter, which unless I'm mistaken only works with handles. I'm a little surprised if that's still the case, though, so perhaps I missed something.

File: dspace-api/src/main/java/org/dspace/app/canvasdimension/CanvasDimensionCLI.java
Comment (mspalti) [reply dspace-api/src/main/java/org/dspace/app/canvasdimension/CanvasDimensionCLI.java]: I agree that this should eventually be updated. Perhaps an issue for 7.3

**Comment by mspalti** [comment]
I think I've addressed all feedback now. The script works with both UUID and handle. If the UUID is provided then an extra dso type parameter is also required so the script knows which service to use when retrieving the dso. e.g.

```
iiif-canvas-dimensions -e user@email.org -i 1086306d-8a51-43c3-98b9-c3b00f49105f -t COLLECTION
```

Example provided in `--help`.

**Comment by tdonohue** [comment]
@mspalti : Apologies, I missed your comment yesterday about UUIDs.  It is possible to resolve an object by ONLY the UUID (no type is necessary) if you use the DSpaceObjectUtils.findDSpaceObject(context, uuid) method: https://github.com/DSpace/DSpace/blob/main/dspace-api/src/main/java/org/dspace/app/util/DSpaceObjectUtilsImpl.java#L37   This is the same manner that our REST API uses to resolve objects based on UUIDs (as UUIDs are unique) when you call the `/api/dso/find?uuid=` endpoint: https://github.com/DSpace/DSpace/blob/main/dspace-server-webapp/src/main/java/org/dspace/app/rest/UUIDLookupRestController.java#L93

That said, as noted in that method, it's not the most performant "lookup" as we literally just try to lookup that UUID for each type of object.  So, while requiring the type is not actually required, it probably will give slightly better performance. 

In this scenario, as you are only doing one object lookup, we probably could drop the requirement for a type...but, honestly, I'm OK either way.  So, your approach in this PR is fine.

I haven't had a chance to give this an updated review, but it does look (at a quick glance) that you've addressed my major concerns.  Will try and get back to this later today if possible.

**Comment by mspalti** [comment]
@tdonohue , that's good to know. It thought there might be such a method. So I think it would be better to drop the "type" parameter. I won't be able to get to that today but it should be a quick fix tomorrow.

**Comment by mspalti** [comment]
OK. This is now ready for another review. I think I've addressed all concerns. Sorry this took a while I'm on the road with a new laptop and bad wifi.

File: dspace-api/src/main/java/org/dspace/iiif/IIIFApiQueryServiceImpl.java
Comment (abollini) [inline dspace-api/src/main/java/org/dspace/iiif/IIIFApiQueryServiceImpl.java:45]: should we just retain the public method?

File: dspace-api/src/main/java/org/dspace/app/canvasdimension/CanvasDimensionCLI.java
Comment (abollini) [reply dspace-api/src/main/java/org/dspace/app/canvasdimension/CanvasDimensionCLI.java]: CurationTask is also something to consider as it can be run as  part of the workflow. We have it in our addon so it is something that we can eventually provide that for 7.3+

**Review by abollini — APPROVED** [review]
changes look good to me. @tdonohue I have reviewed also your question flagging them as resolved where appropriate to speedup your second review round.

**Review by tdonohue — APPROVED** [review]
👍 Reviewed/tested again today. Thanks @mspalti !  This all looks good to me now.  I've verified all my prior feedback was addressed, and re-tested and the commandline tool works well (also works for just UUIDs now).

**Pull request merged**
